src/core/service.c

```diff
@@ -3059,6 +3060,16 @@ static void service_notify_message(Unit *u, pid_t pid, char **tags, FDSet *fds)
                 else
                         log_unit_debug(u, "Got notification message from PID "PID_FMT", but reception only permitted for main PID which is currently not known", pid);
                 return;
+        } else if (s->notify_access == NOTIFY_EXEC && pid != s->main_pid && pid != s->control_pid) {
+                if (s->main_pid != 0 && s->control_pid != 0)
+                        log_unit_warning(u, "Got notification message from PID "PID_FMT", but reception only permitted for main PID "PID_FMT" and control PID "PID_FMT,
+                                         pid, s->main_pid, s->control_pid);
+                else if (s->main_pid != 0 || s->control_pid != 0)
+                        log_unit_warning(u, "Got notification message from PID "PID_FMT", but reception only permitted for %s PID "PID_FMT,
+                                         pid, s->main_pid ? "main" : "control", s->main_pid | s->control_pid);
```

Comment:
we use `?:` pretty frequently... but in this case I'd actually avoid it, as we try to get people to do numeric != 0 checks by spelling out `!= 0`, instead of relying on C's degrade-to-bool feature (we use that only for ptr and bool-like objects, but not for numeric variables)